ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs

```diff
@@ -2933,6 +2990,264 @@ prop_governor_target_active_local_above (MaxTime maxTime) env =
+        govHasOnlyBootstrapPeers =
+          selectGovState Governor.hasOnlyBootstrapPeers events
+
+        isInCleanState :: Signal Bool
+        isInCleanState =
+          fmap (not . Set.null)
+          $ Signal.keyedUntil
+             (\(_, _, ubp, lsj, hp) ->
+               if hp && isInSensitiveState ubp lsj
+                  then Set.singleton ()
+                  else Set.empty
+             )
+             (\(_, _, ubp, lsj, hp) ->
+               if not hp || not (isInSensitiveState ubp lsj)
+                  then Set.singleton ()
+                  else Set.empty
+             )
+             (const False)
+             ((,,,,) <$> govKnownPeers
+                     <*> govTrustedPeers
+                     <*> govUseBootstrapPeers
+                     <*> govLedgerStateJudgement
+                     <*> govHasOnlyBootstrapPeers
+             )
+
+     in signalProperty 20 show
+          (\(b, kp, tp) -> (b && Set.null (Set.difference kp tp)) || not b)
+          ((,,) <$> isInCleanState
+               <*> govKnownPeers
+               <*> govTrustedPeers
+          )
+
+-- | This test checks that if the node is not in a sensitive state it will not
+-- stay connected to _only_ bootstrap peers for too long.
+--
+prop_governor_stops_using_bootstrap_peers :: GovernorMockEnvironment -> Property
+prop_governor_stops_using_bootstrap_peers env =
+    let events = Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . selectPeerSelectionTraceEvents
+               . runGovernorInMockEnvironment
+               $ env
+
+        govUseBootstrapPeers :: Signal UseBootstrapPeers
+        govUseBootstrapPeers =
+          selectGovState Governor.bootstrapPeersFlag events
+
+        govLedgerStateJudgement :: Signal LedgerStateJudgement
+        govLedgerStateJudgement =
+          selectGovState (Governor.ledgerStateJudgement) events
+
+        govLocalRootPeers :: Signal (Set PeerAddr)
+        govLocalRootPeers =
+          selectGovState (LocalRootPeers.keysSet . Governor.localRootPeers) events
+
+        govKnownPeers :: Signal (Set PeerAddr)
+        govKnownPeers =
+          selectGovState (KnownPeers.toSet . Governor.knownPeers) events
+
+        govBootstrapPeers :: Signal (Set PeerAddr)
+        govBootstrapPeers =
+          selectGovState (PublicRootPeers.getBootstrapPeers . Governor.publicRootPeers) events
+
+        keepBootstrapPeersTooLong :: Signal (Set ())
+        keepBootstrapPeersTooLong =
+          Signal.keyedTimeout
+            10 -- seconds
+            (\(knownPeers, localRootPeers, bootstrapPeers, useBootstrapPeers, lsj) ->
+                if not (isInSensitiveState useBootstrapPeers lsj)
+                   then if not (Set.null knownPeers)
+                        && not (Set.null localRootPeers)
+                        && knownPeers `Set.isSubsetOf` bootstrapPeers
+                           then Set.singleton ()
+                           else Set.empty
```

Comment:
We want something like: _there are non-safe peers `=>` we ought to connect to one of them_, isn't it?

Then this looks right to me:
```suggestion
                   then if not (Set.null knownPeers && Set.null nonTrustableLocalRootPeers)
                        || not (knownPeers `Set.isSubsetOf` bootstrapPeers)
                           then Set.singleton ()
                           else Set.empty
```

Reply:
I think you want this which is slightly simpler:
```hs
                   else if not (Set.null knownPeers)
                        && not (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers))
                           then Set.singleton ()
                           else Set.empty
```
This quarantes there's at least one non-trusted peer which unlike in your proposal is not necessarily is a local root peer.  Actually we don't even need to check that `knownPeers` is non-empty, so we can just use:
```hs
                   else if not (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers))
                           then Set.singleton ()
                           else Set.empty
```

Reply:
that doesn't work we have to be careful because the test might run an empty node or there could be an execution trace that can lead to that state. The `isSubsetOf` function is nice because it deals with the empty cases correctly

Reply:
We can pack whole condition to a simple conjunction:
```hs
                if    isInSensitiveState useBootstrapPeers lsj
                   && (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers))
                           then Set.empty
                           else Set.singleton ()
```

Reply:
Yes we can, but that can't be the condition. We need to check if knownPeers is non empty first, then check that known peers contains other peers than the bootstrap peers and trustable local roots. We also need the `localRootPeers /= trustableLocalRootPeers` in the case that the node is configured with all local root peers being trustable and there's no bootstrap peers

Reply:
i.e.

```haskell
                if    not (isInSensitiveState useBootstrapPeers lsj)
                   && not (Set.null knownPeers)
                   && localRootPeers /= trustableLocalRootPeers
                   && knownPeers `Set.isSubsetOf` (bootstrapPeers <> trustableLocalRootPeers)
                   then Set.singleton ()
                   else Set.empty
```

Reply:
That condition assumes that the only non-trustable peers are local roots, so it's not general enough.  It could be that all local roots are trustable, but there are non-trustable ledger peers.  In this case this condition always returnes `True` (or rather `Set.empty`) which is not right.

I think I know where the problem is, we need `thereExistNonTrustablePeer`:
```hs
                if    isInSensitiveState useBootstrapPeers lsj
                   || (   thereExistNonTrustablePeer
                       && (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers)))
                           then Set.empty
                           else Set.singleton ()
                           
```
your condition simplifies the `thereExistNonTrustablePeer` as `localRootPeers /= trustableLocalRootPeers`. Is there a way to make it more general so that it accounts for non-trustable non-local root peers?

Reply:
you are right. I think this works:

```haskell
            if    isInSensitiveState useBootstrapPeers lsj
               && (  not (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers))
                  || Set.null knownPeers)
               then Set.empty
               else Set.singleton ()

```

`thereExistsNonTrustablePeers = not (Set.null (knownPeers Set.\\ bootstrapPeers Set.\\ trustableLocalRootPeers)` but we also have to account for the fact that the knownPeers set might be empty (e.g. quickcheck tries to run an empty node; the node immediately after cleaning its state is empty)